proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,449 @@
+
+      ----------------------------------------------------------------------------
+      ValueError                                Traceback (most recent call last)
+      <ipython-input-16-369cca52e960> in <module>
+            1 # tasks parameters are checked
+      ----> 2 sa(some_param=[Document("Berlin is the capital of Germany.")])
+
+      2 frames
+      /usr/local/lib/python3.8/dist-packages/haystack/nodes/llm/multi_task.py in __call__(self, *args, **kwargs)
+           34         if set(template_dict.keys()) != set(self.prompt_params):
+           35             available_params = set(list(template_dict.keys()) + list(set(kwargs.keys())))
+      ---> 36             raise ValueError(f"Expected prompt params {self.prompt_params} but got {list(available_params)}")
+           37
+           38         template_dict["prompt_template"] = self.prompt_text
+
+      ValueError: Expected prompt params ['documents'] but got ['some_unknown_param']
+  ```
+
+- Pipelines
+
+  Even though we can use PromptNode directly its real power lies in using pipelines and Haystack. For example, we
+  can retrieve documents from the document store using the query and then inject the retrieved documents into documents
+  as a parameter to the selected PromptNode template. For example:
+
+  ``` python
+      from haystack.pipelines import PromptNode
+      top_k = 3
+      query = "Who are the parents of Arya Stark?"
+      retriever = EmbeddingRetriever(...)
+      pn = PromptNode(model_name_or_path="google/flan-t5-base", prompt_template="question-answering")
+
+      pipe = Pipeline()
+      pipe.add_node(component=retriever, name="Retriever", inputs=["Query"])
+      pipe.add_node(component=pn, name="prompt_node", inputs=["Retriever"])
+
+      output = pipe.run(query=query,
+                        params={"Retriever": {"top_k": top_k}},
+                        questions=[query for n in range(0, top_k)],
+                        #documents parameter we need for this task will be automatically populated by the retriever
+                        )
+
+      output["results"]
+  ```
+
+ - However, we are still not utilizing the full power of Haystack pipelines. What if we could use more than
+   one PromptNode in the pipeline? Perhaps we could first retrieve documents from the retriever, pass it
+   to first PromptNode that will generate questions from these documents, and then add a
+   second PromptNode component that will answer those generated questions given the documents as the
+   context. Here is how we can do exactly that:
+
+    ``` python
+        top_k = 3
+        query = "Who are the parents of Arya Stark?"
+        retriever = EmbeddingRetriever(...)
+        model = PromptModel(model_name_or_path="google/flan-t5-small")
+
+        qg = PromptNode(prompt_template="question-generation", prompt_model=model)
+        qa = PromptNode(prompt_template="question-answering", prompt_model=model, output_variable="questions")
```

Comment:
Why are we declaring `output_variable` here and not in the Question generator node `qg`?

Reply:
Yep, proposal mistake. Good catch! Updated, resolving!